crates/storage/nippy-jar/src/writer.rs

```diff
@@ -0,0 +1,341 @@
+            let length = self.offsets_file.metadata()?.len();
+            let bytes_per_offset = 8;
+
+            // Handle non-empty offset file
+            if length > 1 {
+                // first byte is reserved for `bytes_per_offset`, which is 8 initially.
+                let num_offsets = (length - 1) / bytes_per_offset;
+                let new_num_offsets = num_offsets.saturating_sub(remaining_to_prune as u64);
+
+                // If all rows are to be pruned
+                if new_num_offsets <= 1 {
+                    // <= 1 because the one offset would actually be the expected file data size
+                    self.offsets_file.set_len(1)?;
+                    self.data_file.set_len(0)?;
+                } else {
+                    // Calculate the new length for the on-disk offset list
+                    let new_len = 1 + new_num_offsets * bytes_per_offset;
+                    // Seek to the position of the last offset
+                    self.offsets_file
+                        .seek(SeekFrom::Start(new_len.saturating_sub(bytes_per_offset)))?;
+                    // Read the last offset value
+                    let mut last_offset = [0u8; 8];
+                    self.offsets_file.read_exact(&mut last_offset)?;
+                    let last_offset = u64::from_le_bytes(last_offset);
+
+                    // Update the lengths of both the offsets and data files
+                    self.offsets_file.set_len(new_len)?;
+                    self.data_file.set_len(last_offset)?;
+                }
+            } else {
+                // If the offsets file only contains the byte representation
+                self.offsets_file.set_len(1)?;
+                self.data_file.set_len(0)?;
+            }
+        }
+
+        self.offsets_file.seek(SeekFrom::End(0))?;
+        self.data_file.seek(SeekFrom::End(0))?;
+
+        self.jar.rows = self.jar.rows.saturating_sub(num_rows);
+        if self.jar.rows == 0 {
+            self.jar.max_row_size = 0;
+        }
+        self.jar.freeze_config()?;
+
+        Ok(())
+    }
+
+    /// Updates [`NippyJar`] with the new row count and maximum uncompressed row size, while
+    /// resetting internal fields.
+    fn finalize_row(&mut self) {
+        self.jar.max_row_size = self.jar.max_row_size.max(self.uncompressed_row_size);
+        self.jar.rows += 1;
+
+        self.tmp_buf.clear();
+        self.uncompressed_row_size = 0;
+        self.column = 0;
+    }
+
+    /// Commits configuration and offsets to disk. It drains the internal offset list.
+    pub fn commit(&mut self) -> Result<(), NippyJarError> {
+        self.commit_offsets()?;
+
+        // Flushes `max_row_size` and total `rows` to disk.
+        self.jar.freeze_config()?;
+
+        Ok(())
+    }
+
+    /// Flushes offsets to disk.
+    pub(crate) fn commit_offsets(&mut self) -> Result<(), NippyJarError> {
+        // The last offset on disk can be the first offset on `self.offset_list` given how
```

Comment:
why are we sure that only the first offset of `self.offsets` may be on the disk, and not more?

Reply:
`append_column` is the only one that pushes offsets
```rust
                if self.offsets.is_empty() {
                    // Represents the offset of the soon to be appended data column
                    self.offsets.push(self.data_file.stream_position()?);
                }

                self.write_column(value.as_ref())?;

                // Last offset represents the size of the data file if no more data is to be
                // appended. Otherwise, represents the offset of the next data item.
                self.offsets.push(self.data_file.stream_position()?);
```

and commit always flushes its entire list.  unless im missing something, there's no other way that there can be more than the last pushed offset repeated on disk and memory